### ISSUE TITLE
`lizard` does not work

### ISSUE THREAD

**Issue by LecrisUT** [Takishima/cmake-pre-commit-hooks#47]
cmake arguments are not stripped before being passed to the `lizard` executable.

**Comment by Takishima** [comment]
Thanks for reporting this. I will take a look at this issue this week and try to see where the issue lies.

**Comment by Takishima** [comment]
After doing some very basic debugging, I don't seem to be able to reproduce your issues...

I have tried adding a single `print` statement within the `run()` method of `LizardCmd` and using the following configuration:

```yaml
  - repo: local
    hooks:
      - id: lizard
        name: lizard
        entry: ./cmake_pc_hooks/lizard.py
        language: python
        args: [-lcpp, -w]
```

and command line: `pre-commit run --all-files --verbose lizard`, I get the following results:

```
lizard...................................................................Failed
- hook id: lizard
- duration: 0.22s
- exit code: 2

['.editorconfig', '-lcpp', '-w']
['cmake_pc_hooks/lizard.py', '-lcpp', '-w']
['cmake_pc_hooks/__init__.py', '-lcpp', '-w']
...
```

**Comment by Takishima** [comment]
Sorry, my previous answer was beside the point.

Since lizard is does not actually support the CMake arguments, the `lizard` entry point current does not attempt to parse them. You should actually not specify them in the first place for the lizard command.

**Comment by LecrisUT** [comment]
But don't we need the cmake arguments to specify where to build, etc.? Seems it's just that the cmake specific arguments need to be stripped.

**Comment by Takishima** [comment]
Not for the `lizard` hook; that’s not required. It is required for the other hooks though, so specifying the arguments for the other hooks makes sense.

**Comment by Takishima** [comment]
You need to specify the arguments on a hook per hook basis anyway, so simply omitting them in your `pre-commit` configuration files only for `lizard` is what you should be doing in the first place.

**Comment by LecrisUT** [comment]
Indeed it worked. Should be documented though.

The only issue I can see is if it doesn't pick up `configure_file()` generated files.

**Comment by Takishima** [comment]
If the files are within the repository, they should be picked up automatically, but since those are most likely located in the build directory, these will probably not be.

I could see the use of specifying some of the CMake arguments in order to catch those cases, but the issue is that detecting which files within the CMake build directory where the result of a call to `configure_file(...)` is going to be slightly cumbersome...

**Comment by LecrisUT** [comment]
Shouldn't all of them be listed in compile_commands.json, or do they not include them if they are header files?

**Comment by Takishima** [comment]
If those are files that need to be compiled, yes. But I don't think that header files are listed within the `compile_commands.json` file for example.
Plus we'd need to make sure that the `files` and `exclude` keys for each hook from the pre-commit configuration are respected as well which might be a little tricky.

**Comment by LecrisUT** [comment]
To not duplicate I'll move it here:

> > Wouldn't it be more appropriate to just take all the files in `compile_commands.json` and run over all of them there?
> 
> I'll think about it, though I would need to figure out a way to make sure that the `include` and `exclude` directives from pre-commit configuration file also applies to them...

Actually there was either `clang-format` or `clang-tidy` that I've had to include an `exclude` because a header file is actually fortran, not cpp. So I think the `exclude` do apply somehow.

**Comment by Takishima** [comment]
I'll look into it and see what I can come up with.

**Comment by Takishima** [comment]
PR #54 should address your use case. You simply need to add `--read-json-db` on the command line or in the `args: [...]` in the pre-commit configuration file if you want the file list read from the `compile_commands.json` to be appended to the list of files that are considered by each of the hooks.

Note that this is not the default behaviour, because for typical `pre-commit` usage, the files are passed to the hook by the `pre-commit` script itself. since in some cases only a subset of files needs to be processed and not all of them. Adding the new CLI argument will *always* parse the compile database and process the files.

**Comment by LecrisUT** [comment]
> Note that this is not the default behaviour, because for typical pre-commit usage, the files are passed to the hook by the pre-commit script itself.

For `lizard` specifically I think this is what we actually want though right? Unless with pre-commit we can filter to the files that are changed. I will give it a go at some time, but I need to make a test project to check which files are processed.

**Comment by Takishima** [comment]
So, if you install pre-commit as a Git pre-commit hook (no pun intended here...), when you go on to create a Git commit, the hooks will typically only be run on the files that are staged and all the others will be ignored.

**Comment by LecrisUT** [comment]
Hmm, that is indeed ideal. That would indeed be ideal. Maybe we can do some union, but we have to detect the cmake generated files. I think there is some cmake-IDE API support that exposes that (e.g. in CLion, it detects `config.h.in` and uses the parsed version `config.h`).

**Comment by Takishima** [comment]
So the changes in #54 that I will be merging in a bit will allow *always* merging the file lists from the command line and compile database.

Now if we take a look from the perspective of a hook called before performing a Git commit operation for example, as I mentioned above, the hook will receive a list of files that are staged by Git (the others are temporarily stashed typically). So then what we would need to do is something along those lines:

- Make sure that the relevant hooks accept those files; their extensions commonly end in `.in`
- For each of those files, figure out where the generated file is located and run the hook on that one instead of the `.in` file

The main issue I see is that there is no way of knowing where the file gets generated to, without running CMake since `configure_file` calls can literally output the files wherever they want and as far as I can tell, CMake does not keep track of those inside the cache.

**Comment by Takishima** [comment]
One workaround I can see, but which is quite ugly, would be to run a CMake script as an extra step before the hook is invoked where we redefine the `configure_file()` function in order to keep track of where each input-output pair of files are located and store the results in some file that we can then parse when running the hooks.

**Comment by LecrisUT** [comment]
Firstly, sorry I meant intersection, not union. But yeah, the main thing is how to detect `configure_file`. I guess you've already checked [`cmake-file-api`](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html)?

**Comment by Takishima** [comment]
> Firstly, sorry I meant intersection, not union. But yeah, the main thing is how to detect `configure_file`. I guess you've already checked [`cmake-file-api`](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html)?

I had not but now have 😇

There might be a way... I just need to play around a little and see if I can come up with something meaningful.

**Comment by Takishima** [comment]
Ok, so with further experimentation, I don't think we can reliably detect configured files... let me develop a bit.

Assuming that our CMake code contains the following:

```cmake
...
configure_file(${CMAKE_CURRENT_LIST_DIR}/config.h.in config.h)
configure_file(${CMAKE_CURRENT_LIST_DIR}/configured.cpp.in configured.cpp)

add_custom_command(OUTPUT generated.cpp COMMAND touch generated.cpp)
...
```

The `cmake-file-api` would contain the following relevant information:

In `cmakeFiles-v1-XXX.json` (truncated here for presentation purposes)
```json
{
	"inputs" : 
	[
		{
			"path" : "config.h.in"
		},
		{
			"path" : "configured.cpp.in"
		}
	],

}
```

And in `target-XXX-YYY.json` (truncated for presentation purposes)
```json
{
	"sources" : 
	[
		{
			"backtrace" : 1,
			"compileGroupIndex" : 0,
			"path" : "/tmp/build/configured.cpp",
			"sourceGroupIndex" : 0
		},
		{
			"backtrace" : 1,
			"compileGroupIndex" : 0,
			"isGenerated" : true,
			"path" : "/tmp/build/generated.cpp",
			"sourceGroupIndex" : 0
		},
		{
			"backtrace" : 0,
			"isGenerated" : true,
			"path" : "/tmp/build/generated.cpp.rule",
			"sourceGroupIndex" : 1
		}
        ]
}
```

As you can see, the current `cmake-file-api` does not provide a way to detect where the

**Comment by LecrisUT** [comment]
So the `"sources"` is equivalent to `compile_commands.json` as in it does not include headers. Can we find anything helpful in [`cmake-configure-log`](https://cmake.org/cmake/help/latest/manual/cmake-configure-log.7.html) when verbosed? We could also search the `includeDirectories` maybe?

**Comment by Takishima** [comment]
> So the `"sources"` is equivalent to `compile_commands.json` as in it does not include headers. Can we find anything helpful in [`cmake-configure-log`](https://cmake.org/cmake/help/latest/manual/cmake-configure-log.7.html) when verbosed? We could also search the `includeDirectories` maybe?

So as far as I can see, nothing within `cmake-configure-log` would be able to help us.

Manually looking to all possible include directories feels way to cumbersome and error prone if we need to make sure that our search conforms to the one the compiler would make.

**Comment by LecrisUT** [comment]
It's been a while, but have you tried:
```console
$ cmake --trace-expand --trace-format=json-v1 --trace-redirect=log.json
```
(or rather pipe the stderr/stdout directly in python as each line is json object, but it does not add commas to make it all json parseable)

This can fill the gap to detect all `configure_file` commands and handles any `${variable}` expansion. The only special case is when no variable are passed, but that is simply adding `$CMAKE_CURRENT_BINARYDIR` and sourcedir which are evident from the file being called. To ignore third-party, anything from `FETCHCONTENT_BASE_DIR` can be filtered out (`cmake -N -LA` to find custom setting, works on presets also).

**Comment by Takishima** [comment]
> It's been a while, but have you tried:
> 
> ```
> $ cmake --trace-expand --trace-format=json-v1 --trace-redirect=log.json
> ```
> 
> (or rather pipe the stderr/stdout directly in python as each line is json object, but it does not add commas to make it all json parseable)
> 
> This can fill the gap to detect all `configure_file` commands and handles any `${variable}` expansion. The only special case is when no variable are passed, but that is simply adding `$CMAKE_CURRENT_BINARYDIR` and sourcedir which are evident from the file being called. To ignore third-party, anything from `FETCHCONTENT_BASE_DIR` can be filtered out (`cmake -N -LA` to find custom setting, works on presets also).

Did not know this combination of CMake flags. Will take a look at it over the coming days.

**Comment by Takishima** [comment]
Since this issue veered off topic since its beginning, I will be closing it after the last discussion will be resolved using  PR #75.
